### PR TITLE
Only offer the 'add named parameter' refactoring when on the start of the argument being passed in.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/18848

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide

**Comment by CyrusNajmabadi** [comment]
Tagging @MattGertz 

**Customer scenario**

user is in a complex argument to some method (i.e. think a large complex lambda).  During this time, they always see the 'lightbulb' and at any place inside this argument, the lightbulb offers to "use named argument".

This is just very noisy and adds to the feedack we've been getting when lightbulbs are not appropriate contextual

**Bugs this fixes:**

Fixes https://github.com/dotnet/roslyn/issues/18848


**Workarounds, if any**

none

**Risk**

Low.

**Performance impact**

low.

**Is this a regression from a previous update?**

No.  This is a new feature.

**Root cause analysis:**

We often define a code-fix to appear on a node in a tree.  Often this is fine.  However, when those nodes can be arbitrarily large (like an argument to a method), that can lead to a poor experience.

**How was the bug found?**

Internal customer feedback.

File: src/Features/Core/Portable/CodeRefactorings/UseNamedArguments/AbstractUseNamedArgumentsCodeRefactoringProvider.cs [hunk 1]
```diff
@@ -40,6 +45,16 @@ protected abstract class Analyzer<TBaseSyntax, TSyntax, TListSyntax> : IAnalyzer
                     return;
                 }
 
+                // Arguments can be arbitrarily large.  Only offer this feature if the caret is in the
+                // span of the first token so that we don't end up spamming the user with this feature
+                // in places they don't want it.
+
+                var firstToken = argument.GetFirstToken();
```
Comment (sharwell) [inline src/Features/Core/Portable/CodeRefactorings/UseNamedArguments/AbstractUseNamedArgumentsCodeRefactoringProvider.cs]: ❔ What about verbatim string literals?

File: src/EditorFeatures/CSharpTest/CodeActions/UseNamedArguments/UseNamedArgumentsTests.cs [hunk 2]
```diff
@@ -199,5 +199,106 @@ class C : System.Attribute { public C(int arg1) {} public int P { get; set; } }"
...
+
+        [WorkItem(18848, "https://github.com/dotnet/roslyn/issues/18848")]
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseNamedArguments)]
+        public async Task TestMissingNotOnFirstTokenOfArgument2()
+        {
+            await TestMissingAsync(
+@"class C
+{
+    void M(int arg1, int arg2) 
+        => M(1 [||]+ 2, 2);
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/CodeActions/UseNamedArguments/UseNamedArgumentsTests.cs]: 💭 It seems like this change could result in some users believing we removed (or broke) this feature.

File: src/Features/Core/Portable/CodeRefactorings/UseNamedArguments/AbstractUseNamedArgumentsCodeRefactoringProvider.cs [hunk 3]
```diff
@@ -19,17 +19,22 @@ protected interface IAnalyzer
...
+            where TArgumentSyntax : TBaseArgumentSyntax
+            where TArgumentListSyntax : SyntaxNode
         {
             public async Task ComputeRefactoringsAsync(
                 CodeRefactoringContext context, SyntaxNode root, CancellationToken cancellationToken)
             {
                 var document = context.Document;
 
-                var argument = root.FindNode(context.Span).FirstAncestorOrSelf<TBaseSyntax>() as TSyntax;
+                if (context.Span.Length > 0)
```
Comment (sharwell) [inline src/Features/Core/Portable/CodeRefactorings/UseNamedArguments/AbstractUseNamedArgumentsCodeRefactoringProvider.cs]: :memo: I didn't see any new tests for non-empty selections. What was the old behavior?

**Comment by CyrusNajmabadi** [comment]
Algorithm is tweaked slightly.  We now offer the refactoring if your cursor is within the bounds of the argument *and* on the same line that the argument started on.

**Comment by CyrusNajmabadi** [comment]
GOing to merge in once tests pass.  If you want anything else @sharwell, let me know.

**Pull request merged**